apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/HiPerfPublisher.cs

```
@@ -0,0 +1,626 @@
+                {
+                    _hardStopCts.Cancel();
+                    ShutdownIfCompleted();
+                });
+                _taskHelper.Run(async () =>
+                {
+                    await _taskHelper.ConfigureAwaitHideErrors(_shutdownTcs.Task);
+                    registration.Dispose();
+                });
+                ShutdownIfCompleted();
+            }
+            return _shutdownTcs.Task;
+        }
+
+        private void CheckShutdown() => GaxPreconditions.CheckState(!_softStopCts.IsCancellationRequested, "Publisher already shutdown, cannot use");
+
+        private void ShutdownIfCompleted()
+        {
+            // Pre-condition: Must be locked if not a hard stop
+            if (_hardStopCts.IsCancellationRequested || (_softStopCts.IsCancellationRequested && _batchesInFlightCount == 0 && _batchesReady.Count == 0))
+            {
+                if (_shutdownLock.Locked(() => {
+                    var shutdownStarted = _shutdownStarted;
+                    _shutdownStarted = true;
+                    return !shutdownStarted;
+                }))
+                {
+                    // All batches sent and shutdown requested, so signal shutdown completed successfully.
+                    _taskHelper.Run(async () =>
+                    {
+                        if (_shutdown != null)
+                        {
+                            await _taskHelper.ConfigureAwaitHideErrors(_shutdown());
+                        }
+                        if (_hardStopCts.IsCancellationRequested)
+                        {
+                            _shutdownTcs.SetCanceled();
+                        }
+                        else
+                        {
+                            _shutdownTcs.SetResult(0);
+                        }
+                    });
+                }
+            }
+        }
+
+        // Not "async void" to gain standard async exeption handling
```

Except that the task returned is never awaited, so exceptions are just getting swallowed, aren't they? Ditto later for TriggerSend...